### PR TITLE
Add browser testing

### LINKED ISSUE #3: Browser targeted wasm build
Since wasm-bindgen, at the moment, generates *either* nodejs or browser targets, we need to add a browser target, and find some way to test it. Karma?

### PR DESCRIPTION
**Pull request by willemolding**
Closes #3 

Adds browser testing using karma. Can be run using
`npm run test:browser`
from hcid-js directory.

It even uses the same test files as the existing node/mocha tests. How cool is that!

---

What you will notice from the file diff is I had to make some manual changes to the generated bindgen.js and bindgen_bg.js files to make it work in the browser. The first change is using an imported TextEncoder since the util module is not available in the browser. 

The second is changing the loading of the wasm from 
```
const path = require('path').join(__dirname, 'bindgen_bg.wasm');
const bytes = require('fs').readFileSync(path);
```
to 
```
const bytes = require('fs').readFileSync(__dirname + '/bindgen_bg.wasm');
```

Parcel is not able to inline the wasm using the first method which I would consider a bug.

This attempts to test in both Chrome and Firefox. Locally I have Firefox passing but Chrome doesn't allow compiling a wasm binary this large in the main thread. See #9 

---

Without these changes I was not able to get the module running in the browser and this is needed for hClient. Any ideas how I can avoid having to manually modify the generated js files since this breaks the build process? Travis build will not pass because `make test` overwrites the changes.

### CODE REVIEW

**Comment by neonphog** [comment]
@willemolding - I'd start by looking at the wasmbindgen flags - https://github.com/holochain/hcid/blob/master/Makefile#L19 - it's currently set to nodejs, but there's a browser option ... we'd probably need to build both into different directories.

If that doesn't work, we could make use of something like the postinstall package.json hook to apply a patch to the bindgen code.

**Comment by willemolding** [comment]
I have a new solution. This does indeed use the `wasm-bindgen --browser` flag to build a browser compatible version to a separate dir. It also adds the following to the package.json
```
  "browser": {
    "./lib/index.js": "./lib/browser/index.js"
  }
```
which instructs the web bundlers (webpack, parcel, browserify etc) to replace the following file. This works a treat.

There are also some additional kinda hacks. Web support for wasm is pretty iffy at this stage so the safest thing to do is inline the wasm as base64. This is now done in the make file for the browser version only. Unfortunately this seems to be the best way of doing things at the moment while we wait for web bundlers to catch up and standardize.

The other problem is that chrome demands that wasm > 4kB must be compiled and instantiated async. This is taken care of but that does mean the constructor for Encoding now returns a promise. This is fine I think and similar to libsodium_sumo.

All round this was a valuable learning experience but I am very happy with this solution

File: hcid-js/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+- ...
+
+Depending on if you are using the module in node.js or the browser the calling syntax is slightly different. This is because in the browser the WASM must be asynchronously compiled and instantiated to prevent blocking the main thread. As a result all of the constructor returns a promises in the browser but not in node.
+
+```javascript
+
+const publicKey = [...] // UInt8Array of bytes of public key
+
+const enc = new Encoding('hcs') // node.js
+const enc = await new Encoding('hcs') // browser
```
Comment (neonphog) [inline hcid-js/README.md]: these need `0`s ... `new Encoding('hcs0')` we've got a version bit in there we can update when we get a 5 bit reed-solomon encoder in rust.

**Comment by neonphog** [comment]
huge thread about chromeheadless on osx for the ci failures... not sure if it's helpful: https://github.com/karma-runner/karma-chrome-launcher/issues/154

**Review by neonphog — APPROVED** [review]
:guitar: this is awesome - readme fix and sort out ci, and should be good to go :+1:

**Comment by willemolding** [comment]
For future reference using:
```
addons:
  chrome: stable
```
in the .travis.toml file fixed the problem

**Pull request merged**